056-zk-kraft-migration.md

```diff
@@ -0,0 +1,254 @@
+The FSM moves back to the `WaitBrokersInMigration`, so actually at the beginning of the Phase 2.
+
+If the user doesn't rollback the migration, next step is to move brokers and controllers out of migration and using KRaft only in Phase 3/4.
+
+#### Phase 3/4: Migrating brokers and controllers to full KRaft
+
+In the `DualWrite` state, the operator expects the user desire to finalize the migration, having a full KRaft based cluster and ZooKeeper out of the picture.
+The user applies the `strimzi.io/kraft-migration: post-migration` annotation to the `Kafka` custom resource.
+The operator updates the brokers' configuration by removing the ZooKeeper migration flag and connection to the ZooKeeper ensemble.
+It also has to replace the `broker.id` with `node.id`.
+
+```shell
+# replace broker.id with node.id
+node.id=0
+
+# Remove the migration enabled flag
+# zookeeper.metadata.migration.enable=true
+
+# Remove ZooKeeper client configuration
+# zookeeper.connect=<connection-to-zookeeper>
+
+# Don't set the IBP, KRaft uses "metadata.version" feature flag
+# inter.broker.protocol.version=3.5
+```
+
+The operator rolls the brokers and waits for them to be ready again.
+They are now connected to the KRaft controllers quorum only and not to the ZooKeeper ensemble anymore.
+
+When brokers are rolled and ready again, the operator updates the controllers' configuration by removing the ZooKeeper migration flag and connection to the ZooKeeper ensemble.
+
+```shell
+# Disable the migration
+# zookeeper.metadata.migration.enable=true
+
+# Remove ZooKeeper client configuration
+# zookeeper.connect=<connection-to-zookeeper>
+```
+
+Finally, the operator rolls the controllers and wait for them to be ready again.
+They are now disconnected from the ZooKeeper ensemble which is totally out of the picture.
+The FSM moves to the `FullKRaft` state and ZooKeeper is not used anymore.
+
+Before completing the migration, the `DualWrite` state is also the last opportunity for the user to rollback the migration process by applying the `strimzi.io/kraft-migration: pre-migration` on the `Kafka` custom resource.
+In this case, the operator configures the brokers with no ZooKeeper migration flag, and no KRaft controllers quorum connection information.
+Finally, the operator rolls the brokers and wait for them to be ready again.
+The FSM moves back to the `WaitBrokersInMigration`, so actually at the beginning of the Phase 2.
+
+#### Phase 5: Removing ZooKeeper
+
+At this point the ZooKeeper ensemble is out of the picture, still running but not used anymore.
+In the `FullKRaft` state, the operator expects the user to finalize the migration by removing ZooKeeper.
+The user removes the `spec.zookeeper` on the `Kafka` custom resource so that the operator can delete the ZooKeeper nodes.
+The FSM moves back to a `NoMigration` state (or null) and the operator upgrades `Kafka.status.migrationState` field accordingly.
+
```

Comment:
Overall the descriptions of the phases make sense, but I think they could be written in a more structured way which would make it more easily understood. 

Each phase could be described using the same template. You could use subsections, or start each paragraph with bold. 
1. The **triggering action** (usually this is the user changing the annotation). 
2. A number of paragraph covering all the **preconditions** that are verified before actions by the operator, and what happens if they're violated. This might include the current state of the migration in the Kafka CR, for example.
3. The **actions** that the operator does (if the preconditions were met), along with things which might be observable during the actions (e.g. changes to Kafka metrics) 
4. The Kafka CR **status changes** that the operator makes when the action is completed (successfully and unsuccessfully).

 You're also to entirely clear with things like `WaitBrokersInMigration` whether these are CR states, or merely states which exist within Kafka itself.

Reply:
> Each phase could be described using the same template.

I think this is a really great idea. It could make everything more structured and understandable. I will work on it.

> You're also to entirely clear with things like `WaitBrokersInMigration` whether these are CR states, or merely states which exist within Kafka itself.

These are states in the migration FSM, so states of the `KRaftMigrationManager` which will be a component within the operator. These states are persisted into the `Kafka.status.migrationState` field.

Reply:
I updated both the FSM graph and the proposal making each phase more structured.
Only doubt is that I just used a divider when in a phase you can have a different transitions across the graph (because different preconditions, triggers).